### ISSUE TITLE
Question - Bug reporting

### ISSUE THREAD

**Issue by LaurenceLeo** [hezmondo/Jinn#134]
At the moment when an error occurs we are reliant upon the users to inform us of what happened, where, and when. I was wondering if it would be possible for us to add the ability for a user to report a bug when it occurs. This way we can get the information we need as quickly as possible whilst relying upon the users competency as little as possible. 

I was thinking that this could work in the following way:

Error is raised in Python/Jinn
Error is caught by bugreporter
User is given the option to report the bug
Log is formatted into an email and sent to "guavasystemslimited@gmail.com"
(using the users SMTP settings, if they don't have any then this cannot work)

Does this seem like a good/possible idea? Is there a better way to do this?

**Comment by jnbarchan** [comment]
Yes, we could do something like.  My company software has this in it, but only triggered by serious errors.  I _think_ I've received an email from such a fault about once in ten years :)

* Problem in Jinn is _when_ should this be triggered.  Jinn. tends to do stuff like raise an error when the user clicks to cancel a dialog, or picks a bad filename to open, etc.  We don't want to hear about those!

* Situation does not cover where nothing has gone "wrong" in the software (it hasn't errored) but it is not "behaving right".  What about that?

* SMTP settings are (now) stored in the database, not the config file.  Any error occurring _before_ successful connection to the database (e.g. during start-up or upgrade) will not have the details to email.  Just saying....

* Although maybe better than nothing, I find the worst part is that the error is often useless without know the context/what the user was doing/what the user's data is.  My company's applications _begs_ the user to describe what they were up to when the dialog comes up inviting them to submit the details by email, and 90% of the time the user ignores that and just clicks **Send** without telling us anything!

But may be better than nothing anyway.

**Comment by LaurenceLeo** [comment]
Thanks for your response, it's interesting to see your experience from actual use of a similar concept. 
I agree that there are indeed some flaws in this approach, but as you say it is better than nothing.

I do actually want to catch those small error messages you mentioned in your first point. That way I can deal with them quickly and push fixes without having to bother you or Hes with small problems. 

I do find your last point very interesting, it seems as though no matter what we do the user will manage to be a useless as possible!! :D

**Comment by jnbarchan** [comment]
> I do actually want to catch those small error messages you mentioned in your first point

Then I think you/someone will need to go through every possible "error" in Jinn code and sort out/re-code which are "programming logic errors" vs "normal/anticipated errors" as I indicated.  Apart from "Cancel"-type stuff, Jinn allows/issues a lot of errors on perfectly normal things: for instance, we do *not* want the user to be prompted that "Jinn has encountered an error, please submit the email" when all they have done is pick a bad filename or something....
